pkg/cmd/extension/manager.go

```diff
@@ -320,41 +381,82 @@ func (m *Manager) Upgrade(name string, force bool, stdout, stderr io.Writer) err
 	someUpgraded := false
 	for _, f := range exts {
 		if name == "" {
-			fmt.Fprintf(stdout, "[%s]: ", f.Name())
+			fmt.Fprintf(m.io.Out, "[%s]: ", f.Name())
 		} else if f.Name() != name {
 			continue
 		}
 
 		if f.IsLocal() {
 			if name == "" {
-				fmt.Fprintf(stdout, "%s\n", localExtensionUpgradeError)
+				fmt.Fprintf(m.io.Out, "%s\n", localExtensionUpgradeError)
 			} else {
 				err = localExtensionUpgradeError
 			}
 			continue
 		}
 
-		var cmds []*exec.Cmd
-		dir := filepath.Dir(f.Path())
-		if force {
-			fetchCmd := m.newCommand(exe, "-C", dir, "--git-dir="+filepath.Join(dir, ".git"), "fetch", "origin", "HEAD")
-			resetCmd := m.newCommand(exe, "-C", dir, "--git-dir="+filepath.Join(dir, ".git"), "reset", "--hard", "origin/HEAD")
-			cmds = []*exec.Cmd{fetchCmd, resetCmd}
-		} else {
-			pullCmd := m.newCommand(exe, "-C", dir, "--git-dir="+filepath.Join(dir, ".git"), "pull", "--ff-only")
-			cmds = []*exec.Cmd{pullCmd}
+		binManifestPath := filepath.Join(filepath.Dir(f.Path()), manifestName)
+		if _, e := os.Stat(binManifestPath); e == nil {
+			err = m.upgradeBin(f)
+			someUpgraded = true
```

Comment:
Should we check that the error is `nil` first before setting this to `true`?

Reply:
my reading of the code (which might be off base) is that `someUpgraded` is capturing whether or not we matched on a name, not whether we successfully upgraded. At least, that's how it functions prior to the PR and I tried to mirror it.

As you suggested above I think the error handling in this loop needs to be rethought either way.